tests/unittests/general/webfront_test.py

```diff
@@ -16,33 +16,48 @@
+    def test_authenticate_ldap_should_return_account_when_ldap_says_yes(self):
         with patch("nav.web.ldapauth.available", new=True):
             with patch("nav.web.ldapauth.authenticate", return_value=True):
                 assert auth.authenticate_ldap('knight', 'shrubbery') == LDAP_ACCOUNT
 
-    def test_authenticate_should_return_false_when_ldap_says_no(self):
+    def test_authenticate_ldap_should_return_false_when_ldap_says_no(self):
         with patch("nav.web.ldapauth.available", new=True):
             with patch("nav.web.ldapauth.authenticate", return_value=False):
-                assert auth.authenticate_ldap('knight', 'shrubbery') is None
+                assert auth.authenticate_ldap('knight', 'shrubbery') is False
 
-    def test_authenticate_should_fallback_when_ldap_is_disabled(self):
+    def test_authenticate_ldap_should_fallback_when_ldap_is_disabled(self):
         with patch("nav.web.ldapauth.available", new=False):
             assert auth.authenticate_ldap('knight', 'shrubbery') is None
 
 
-@patch("nav.web.auth.Account.save", new=MagicMock(return_value=True))
 @patch("nav.web.auth.Account.objects.get",
        new=MagicMock(return_value=PLAIN_ACCOUNT))
 @patch("nav.web.ldapauth.available", new=False)
 class TestNormalAuthenticate(object):
-    def test_authenticate_should_return_account_when_password_is_ok(self):
+    def test_authenticate_account_should_return_account_when_password_is_ok(self):
         with patch("nav.web.auth.Account.check_password", return_value=True):
-            assert auth.authenticate('knight', 'shrubbery') == PLAIN_ACCOUNT
+            assert auth.authenticate_account('knight', 'shrubbery') == PLAIN_ACCOUNT
 
-    def test_authenticate_should_return_false_when_ldap_says_no(self):
+    def test_authenticate_account_should_return_falsey_when_password_is_wrong(self):
         with patch("nav.web.auth.Account.check_password", return_value=False):
-            assert not auth.authenticate('knight', 'rabbit')
+            assert not auth.authenticate_account('knight', 'rabbit')
+
+
+def test_authenticate_should_return_ldap_account_when_ldap_user_exists():
+    with patch("nav.web.ldapauth.available", new=True):
+        with patch("nav.web.ldapauth.authenticate", return_value=True):
+            with patch("nav.web.auth.Account.save", new=MagicMock(return_value=True)):
+                with patch("nav.web.auth.Account.check_password", return_value=True):
+                    with patch("nav.web.auth.Account.objects.get",
+                               new=MagicMock(return_value=LDAP_ACCOUNT)):
+                        assert auth.authenticate('knight', 'shrubbery') == LDAP_ACCOUNT
+
+
+def test_authenticate_should_return_none_when_ldap_user_does_not_exist():
+    with patch("nav.web.ldapauth.available", new=True):
+        with patch("nav.web.ldapauth.authenticate", return_value=False):
+            assert auth.authenticate('knight', 'shrubbery') == None
 
```

Comment:
given that `authenticate_account` and `authenticate_ldap` has good unit tests already, the tests for `authenticate` should probably mock these, rather than all the mocking required to by them. 

Using `patch(..., new=MagicMock(foo))` is a trick to prevent patch from adding the mock object as a new parameter to the function when using patch as a decorator. When using patch as a context manager it is cleaner to use `patch(..., foo)`

Reply:
"Simplify" mocks the authenticate_*-methods.